File: Leetcode/239. Sliding Window Maximum.py
import heapq


class Solution:
    # using a prior queue to store the max value
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        q = [(-nums[i], i) for i in range(k)]
        heapq.heapify(q)

        res = [-q[0][0]]

        for i in range(k, len(nums)):
            heapq.heappush(q, (-nums[i], i))
            while q[0][1] <= i-k:
                heapq.heappop(q)
            res.append(-q[0][0])

        return res


class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        q = collections.deque()
        res = []

        for i in range(k):
            while q and nums[i] > nums[q[-1]]:
                q.pop()
            q.append(i)

        res.append(nums[q[0]])

        for i in range(k, len(nums)):
            while q and nums[i] > nums[q[-1]]:
                q.pop()
            q.append(i)
            while q[0] <= i - k:
                q.popleft()

            res.append(nums[q[0]])

        return res


input = [1, 3, -1, -3, 5, 3, 6, 7]

res = Solution().maxSlidingWindow(input, 3)
print(res)

File: Leetcode/239. Sliding Window Maximum.ts
// brute force
function maxSlidingWindow(nums: number[], k: number): number[] {
  const res = [];
  for (let i = 0, j = k - 1; j < nums.length; i++, j++) {
    res.push(Math.max(...nums.slice(i, j + 1)));
  }

  return res;
}

let input = [9, 10, 9, -7, -4, -8, 2, -6];
let k = 5;

console.log(maxSlidingWindow(input, k));

// using a queue to store the index of max value
function maxSlidingWindow(nums: number[], k: number): number[] {
  const res = [];
  const queue = []; // store the index

  for (let i = 0; i < k; i++) {
    while (queue.length && nums[i] > nums[queue[queue.length - 1]]) {
      queue.pop();
    }
    queue.push(i);
  }

  res.push(nums[queue[0]]);

  for (let i = k; i < nums.length; i++) {
    // Add the new index to the queue
    while (queue.length && nums[i] > nums[queue[queue.length - 1]]) {
      queue.pop();
    }
    queue.push(i);

    // check whether the max value index is in the current slide window
    while (queue[0] <= i - k) {
      queue.shift();
    }

    res.push(nums[queue[0]]);
  }

  return res;
}
